apps/settings/js/modules/addon_manager.js

```diff
@@ -105,6 +108,108 @@ define(function(require) {
     });
   }
 
+  function renameAddon(addon, name) {
+    if (!name) {
+      return Promise.reject('no name given');
+    }
+
+    if (!isAddon(addon)) {
+      return Promise.reject('not an addon');
+    }
+
+    var manifest = getManifest(addon);
+    if (name === manifest.name) {
+      return Promise.reject('name is unchanged');
+    }
+
+    // To rename an addon we are creating a new DOMApplication that will have
+    // an updated name. Other manifest attributes are preserved.
+    return new Promise(function(resolve, reject) {
+      // Export an addon as a blob.
+      addon.export().then(function(blob) {
+        // Copy over manifest object.
+        var newManifest = Object.assign({}, manifest);
+        newManifest.name = name;
+
+        deleteAddon(addon).then(function() {
+          blobToArrayBuffer(blob).then(function(arrayBuffer) {
```

Comment:
As we are inside a promise chain, I would suggest to avoid nested functions if possible for readability.